{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, pysam, matplotlib.pyplot as plt, numpy as np, pandas as pd, seaborn as sns\n",
    "from glob import glob\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " lfermentum NZ_BJLV01000146.1     \r"
     ]
    }
   ],
   "source": [
    "def get_motifs(ref, seq, motif2pos):\n",
    "    \"\"\"Return position of motifs in sequence\"\"\"\n",
    "    motifs = {m: set() for m in motif2pos}\n",
    "    motifs_lens = set(len(m) for m in motifs)#; print(motifs_lens)\n",
    "    # only count + strand\n",
    "    strand = \"+\"\n",
    "    for s in range(len(seq)-min(motifs_lens)):\n",
    "        for l in motifs_lens:\n",
    "            m = seq[s:s+l]\n",
    "            if m in motifs:\n",
    "                pos = \"%s:%s%s\"%(ref, s+motif2pos[m], strand)\n",
    "                motifs[m].add(pos)\n",
    "    return motifs\n",
    "\n",
    "def fasta2data(fasta, motif2pos):\n",
    "    \"\"\"Return \"\"\"\n",
    "    s2d, s2c = {}, {}\n",
    "    for r in SeqIO.parse(fasta, \"fasta\"):\n",
    "        ref = r.name\n",
    "        seq = str(r.seq)\n",
    "        family, species = r.description.split()[1:3]\n",
    "        spname = \"%s%s\"%(family.lower()[0], species)\n",
    "        sys.stderr.write(\" %s %s     \\r\"%(spname, ref))\n",
    "        # get motifs\n",
    "        motifs = get_motifs(ref, seq, motif2pos)\n",
    "        # update species motifs\n",
    "        if spname in s2d: \n",
    "            for m in motif2pos: \n",
    "                s2d[spname][m] = s2d[spname][m].union(motifs[m])\n",
    "        # all motifs are fresh here\n",
    "        else: \n",
    "            s2d[spname] = motifs #{m: set() for m in motif2pos}\n",
    "            s2c[spname] = []\n",
    "        s2c[spname].append(ref)\n",
    "    return s2d, s2c\n",
    "\n",
    "fasta = \"/home/lpryszcz/cluster/dna_mods/ref/mock_community.fa\"\n",
    "\n",
    "mod2mers = {\"6mA\": ((2, \"GATC\"), ), \"5mC\": ((2, \"CCAGG\"), (2, \"CCTGG\"), (1, \"CG\"))}; mod2mers\n",
    "motif2pos = {m: p for mod in mod2mers for p, m in mod2mers[mod]}; motif2pos\n",
    "\n",
    "species2motifs, species2refs = fasta2data(fasta, motif2pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>strand</th>\n",
       "      <th>mod</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam depth</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam basecall_accuracy</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam mod_frequency</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam median_mod_prob</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam depth</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam basecall_accuracy</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam mod_frequency</th>\n",
       "      <th>modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam median_mod_prob</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>244</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>444</td>\n",
       "      <td>0.910</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.833</td>\n",
       "      <td>120</td>\n",
       "      <td>0.958</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.600</td>\n",
       "      <td>NC_000913.3:244-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>420</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>5mC</td>\n",
       "      <td>464</td>\n",
       "      <td>0.978</td>\n",
       "      <td>0.713</td>\n",
       "      <td>0.967</td>\n",
       "      <td>132</td>\n",
       "      <td>0.962</td>\n",
       "      <td>0.644</td>\n",
       "      <td>0.967</td>\n",
       "      <td>NC_000913.3:420+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>422</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>351</td>\n",
       "      <td>0.604</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.833</td>\n",
       "      <td>103</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.369</td>\n",
       "      <td>0.867</td>\n",
       "      <td>NC_000913.3:422-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>423</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>455</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.683</td>\n",
       "      <td>123</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.057</td>\n",
       "      <td>0.767</td>\n",
       "      <td>NC_000913.3:423-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>475</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>5mC</td>\n",
       "      <td>432</td>\n",
       "      <td>0.942</td>\n",
       "      <td>0.664</td>\n",
       "      <td>0.933</td>\n",
       "      <td>126</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.635</td>\n",
       "      <td>0.967</td>\n",
       "      <td>NC_000913.3:475+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr  pos ref_base strand  mod  \\\n",
       "0  NC_000913.3  244        C      -  5mC   \n",
       "1  NC_000913.3  420        C      +  5mC   \n",
       "2  NC_000913.3  422        C      -  5mC   \n",
       "3  NC_000913.3  423        C      -  5mC   \n",
       "4  NC_000913.3  475        C      +  5mC   \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam depth  \\\n",
       "0                                                444                                  \n",
       "1                                                464                                  \n",
       "2                                                351                                  \n",
       "3                                                455                                  \n",
       "4                                                432                                  \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam basecall_accuracy  \\\n",
       "0                                              0.910                                              \n",
       "1                                              0.978                                              \n",
       "2                                              0.604                                              \n",
       "3                                              0.993                                              \n",
       "4                                              0.942                                              \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam mod_frequency  \\\n",
       "0                                              0.014                                          \n",
       "1                                              0.713                                          \n",
       "2                                              0.328                                          \n",
       "3                                              0.022                                          \n",
       "4                                              0.664                                          \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_1D-Ecoli-run_FAF05145.bam median_mod_prob  \\\n",
       "0                                              0.833                                            \n",
       "1                                              0.967                                            \n",
       "2                                              0.833                                            \n",
       "3                                              0.683                                            \n",
       "4                                              0.933                                            \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam depth  \\\n",
       "0                                                120                                  \n",
       "1                                                132                                  \n",
       "2                                                103                                  \n",
       "3                                                123                                  \n",
       "4                                                126                                  \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam basecall_accuracy  \\\n",
       "0                                              0.958                                              \n",
       "1                                              0.962                                              \n",
       "2                                              0.621                                              \n",
       "3                                              0.984                                              \n",
       "4                                              0.937                                              \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam mod_frequency  \\\n",
       "0                                              0.050                                          \n",
       "1                                              0.644                                          \n",
       "2                                              0.369                                          \n",
       "3                                              0.057                                          \n",
       "4                                              0.635                                          \n",
       "\n",
       "   modPhred/PRJEB22772/minimap2/MARC_ZFscreens_R9.4_2D-Ecoli-run_FAF05711.bam median_mod_prob  \\\n",
       "0                                              0.600                                            \n",
       "1                                              0.967                                            \n",
       "2                                              0.867                                            \n",
       "3                                              0.767                                            \n",
       "4                                              0.967                                            \n",
       "\n",
       "                ref  \n",
       "0  NC_000913.3:244-  \n",
       "1  NC_000913.3:420+  \n",
       "2  NC_000913.3:422-  \n",
       "3  NC_000913.3:423-  \n",
       "4  NC_000913.3:475+  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# #/home/lpryszcz/cluster/dna_mods/ecoli/modPhred.2020/PRJNA477598/\n",
    "fn = \"/home/lpryszcz/cluster/dna_mods/ecoli/modPhred.2020/PRJEB22772/mod.gz\" # PRJNA477598 PRJEB22772\n",
    "df = pd.read_csv(fn, sep=\"\\t\", skiprows=16)#; df.head() \n",
    "# get ref as chr:pos + strand\n",
    "df[\"ref\"] = df['chr'].astype(str) + \":\" + df['pos'].astype(str) + df['strand']; df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ecoli 904.8964473568217 38897 29165 0.6316899760907011 0.42358745071146925 68062 0.9979606776824932 0.9966672221296451 0.9978508844437097 0.0011966792870675015\n",
      "ecoli 250.80205956351296 38495 29869 0.5925278088063384 0.4410661220663565 68364 0.996705710102489 0.9988335277453758 0.9988427839312283 0.004253257707047144\n"
     ]
    }
   ],
   "source": [
    "min_freq = 0.05\n",
    "spname = \"ecoli\"\n",
    "for spcol in filter(lambda c: c.endswith(\"mod_frequency\"), df.columns):\n",
    "    #spname = spcol.split(\"_\")[1]\n",
    "    if spname not in species2motifs: \n",
    "        print(spname, \"is missing\")\n",
    "        continue\n",
    "    _df = df[(df[spcol]>=min_freq)&(df[\"chr\"].isin(species2refs[spname]))]\n",
    "    data = []\n",
    "    # depth 6mA\t5mC\tmean 6mA freq\tmean m5C freq\tpositions\n",
    "    data.append(2*_df[spcol.replace(\"mod_frequency\", \"depth\")].mean())\n",
    "    # get number of every modification\n",
    "    for mod in mod2mers.keys():\n",
    "        data.append((_df[\"mod\"]==mod).sum())\n",
    "    # get avg mod_freq\n",
    "    for mod in mod2mers.keys():\n",
    "        data.append(_df.loc[_df[\"mod\"]==mod, spcol].mean())\n",
    "    # store total number of modifications\n",
    "    data.append(len(_df))    \n",
    "    # get each motif penetrance\n",
    "    for m, mpos in species2motifs[spname].items():\n",
    "        common = mpos.intersection(_df[\"ref\"].to_numpy())\n",
    "        data.append(len(common)/len(mpos))\n",
    "    print(spname, *data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>strand</th>\n",
       "      <th>mod</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam depth</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam basecall_accuracy</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam mod_frequency</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam median_mod_prob</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode01_paeruginosa_pass.bam depth</th>\n",
       "      <th>...</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/barcode08_scerevisiae_pass.bam median_mod_prob</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam depth</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam basecall_accuracy</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam mod_frequency</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam median_mod_prob</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam depth</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam basecall_accuracy</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam mod_frequency</th>\n",
       "      <th>modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam median_mod_prob</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>23</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.516</td>\n",
       "      <td>13</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000913.3:23-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>24</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.710</td>\n",
       "      <td>14</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000913.3:24-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>35</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>0.938</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.774</td>\n",
       "      <td>NC_000913.3:35-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>123</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>0.944</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.806</td>\n",
       "      <td>NC_000913.3:123-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000913.3</td>\n",
       "      <td>153</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>5mC</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.875</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19</td>\n",
       "      <td>0.947</td>\n",
       "      <td>0.105</td>\n",
       "      <td>0.919</td>\n",
       "      <td>NC_000913.3:153-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr  pos ref_base strand  mod  \\\n",
       "0  NC_000913.3   23        C      -  5mC   \n",
       "1  NC_000913.3   24        C      -  5mC   \n",
       "2  NC_000913.3   35        A      -  5mC   \n",
       "3  NC_000913.3  123        T      -  5mC   \n",
       "4  NC_000913.3  153        C      -  5mC   \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam depth  \\\n",
       "0                                                  0                    \n",
       "1                                                  0                    \n",
       "2                                                  0                    \n",
       "3                                                  0                    \n",
       "4                                                  0                    \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam basecall_accuracy  \\\n",
       "0                                                NaN                                \n",
       "1                                                NaN                                \n",
       "2                                                NaN                                \n",
       "3                                                NaN                                \n",
       "4                                                NaN                                \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam mod_frequency  \\\n",
       "0                                                NaN                            \n",
       "1                                                NaN                            \n",
       "2                                                NaN                            \n",
       "3                                                NaN                            \n",
       "4                                                NaN                            \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode01_cneoformans_pass.bam median_mod_prob  \\\n",
       "0                                                NaN                              \n",
       "1                                                NaN                              \n",
       "2                                                NaN                              \n",
       "3                                                NaN                              \n",
       "4                                                NaN                              \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode01_paeruginosa_pass.bam depth  ...  \\\n",
       "0                                                  0                   ...   \n",
       "1                                                  0                   ...   \n",
       "2                                                  0                   ...   \n",
       "3                                                  0                   ...   \n",
       "4                                                  0                   ...   \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/barcode08_scerevisiae_pass.bam median_mod_prob  \\\n",
       "0                                                NaN                              \n",
       "1                                                NaN                              \n",
       "2                                                NaN                              \n",
       "3                                                NaN                              \n",
       "4                                                NaN                              \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam depth  \\\n",
       "0                                                  6      \n",
       "1                                                  7      \n",
       "2                                                  8      \n",
       "3                                                  8      \n",
       "4                                                  8      \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam basecall_accuracy  \\\n",
       "0                                              1.000                  \n",
       "1                                              1.000                  \n",
       "2                                              1.000                  \n",
       "3                                              1.000                  \n",
       "4                                              0.875                  \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam mod_frequency  \\\n",
       "0                                              0.167              \n",
       "1                                              0.143              \n",
       "2                                              0.000              \n",
       "3                                              0.000              \n",
       "4                                              0.000              \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_chiu_r9.bam median_mod_prob  \\\n",
       "0                                              0.516                \n",
       "1                                              0.710                \n",
       "2                                                NaN                \n",
       "3                                                NaN                \n",
       "4                                                NaN                \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam depth  \\\n",
       "0                                                 13       \n",
       "1                                                 14       \n",
       "2                                                 16       \n",
       "3                                                 18       \n",
       "4                                                 19       \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam basecall_accuracy  \\\n",
       "0                                              1.000                   \n",
       "1                                              1.000                   \n",
       "2                                              0.938                   \n",
       "3                                              0.944                   \n",
       "4                                              0.947                   \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam mod_frequency  \\\n",
       "0                                              0.000               \n",
       "1                                              0.000               \n",
       "2                                              0.062               \n",
       "3                                              0.056               \n",
       "4                                              0.105               \n",
       "\n",
       "   modPhred/PRJNA477598/minimap2/nasa_mason_r9.bam median_mod_prob  \\\n",
       "0                                                NaN                 \n",
       "1                                                NaN                 \n",
       "2                                              0.774                 \n",
       "3                                              0.806                 \n",
       "4                                              0.919                 \n",
       "\n",
       "                ref  \n",
       "0   NC_000913.3:23-  \n",
       "1   NC_000913.3:24-  \n",
       "2   NC_000913.3:35-  \n",
       "3  NC_000913.3:123-  \n",
       "4  NC_000913.3:153-  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# #/home/lpryszcz/cluster/dna_mods/ecoli/modPhred.2020/PRJNA477598/\n",
    "fn = \"/home/lpryszcz/cluster/dna_mods/ecoli/modPhred.2020/PRJNA477598/mod.gz\" # PRJNA477598 PRJEB22772\n",
    "df = pd.read_csv(fn, sep=\"\\t\", skiprows=16)#; df.head() \n",
    "# get ref as chr:pos + strand\n",
    "df[\"ref\"] = df['chr'].astype(str) + \":\" + df['pos'].astype(str) + df['strand']; df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cneoformans 65.78276113133576 17204 72071 0.13394576842594746 0.10239215495830499 89275 0.1017274472168906 0.0021475580187045374 0.001169751599807335 0.03904561765230666\n",
      "paeruginosa 147.31385435637554 65784 9990 0.26019145992946613 0.07769729729729731 75774 0.776373881015684 0.0006533812479581836 0.0007201880069744522 0.00339133192333598\n",
      "ecoli 346.96027108433736 36912 29488 0.6344618010403121 0.37014775501899083 66400 0.9391863626856306 0.9385102482919513 0.9378409654488346 0.004368600288933168\n",
      "ecoli 45.61268275073702 39511 43594 0.592452582825036 0.26199488461714915 83105 0.9376699435264589 0.928178636893851 0.9332121011737478 0.013627726049833762\n",
      "lfermentum nan 0 0 nan nan 0 0.0 0.0 0.0 0.0\n",
      "efaecalis 374.61111966733404 2052 4441 0.1252982456140351 0.08072798919162352 6493 0.13530685920577618 0.0 0.0 0.019455486773531608\n",
      "lmonocytogenes 195.55796120884077 1982 4669 0.12580575176589304 0.0788935532233883 6651 0.15637065637065636 0.0010384215991692627 0.001076426264800861 0.016851557479622553\n",
      "bsubtilis 171.59630984571336 14132 11016 0.16467258703651286 0.17000962236746547 25148 0.27367660945397987 0.48431618569636137 0.4987980769230769 0.010204378007338633\n",
      "senterica 73.40781530078154 37436 32299 0.6323762421198845 0.3194278460633456 69735 0.8810900082576383 0.8416283084004603 0.8255729090643702 0.003907230168118436\n",
      "saureus 243.84596057209123 79 10269 0.08877215189873419 0.14987213944882657 10348 0.0029222676797194622 0.0 0.0 0.04608685975015998\n",
      "scerevisiae 87.23990449496523 973 8660 0.11636176772867421 0.09157321016166282 9633 0.013924734409699298 0.0 0.00017866714311238162 0.013507732166951381\n",
      "chiu is missing\n",
      "mason is missing\n"
     ]
    }
   ],
   "source": [
    "for spcol in filter(lambda c: c.endswith(\"mod_frequency\"), df.columns):\n",
    "    spname = spcol.split(\"_\")[1]\n",
    "    if spname not in species2motifs: \n",
    "        print(spname, \"is missing\")\n",
    "        continue\n",
    "    _df = df[(df[spcol]>=min_freq)&(df[\"chr\"].isin(species2refs[spname]))]\n",
    "    data = []\n",
    "    # depth 6mA\t5mC\tmean 6mA freq\tmean m5C freq\tpositions\n",
    "    data.append(2*_df[spcol.replace(\"mod_frequency\", \"depth\")].mean())\n",
    "    # get number of every modification\n",
    "    for mod in mod2mers.keys():\n",
    "        data.append((_df[\"mod\"]==mod).sum())\n",
    "    # get avg mod_freq\n",
    "    for mod in mod2mers.keys():\n",
    "        data.append(_df.loc[_df[\"mod\"]==mod, spcol].mean())    \n",
    "    # store total number of modifications\n",
    "    data.append(len(_df))\n",
    "    # get each motif penetrance\n",
    "    for m, mpos in species2motifs[spname].items():\n",
    "        common = mpos.intersection(_df[\"ref\"].to_numpy())\n",
    "        data.append(len(common)/len(mpos))\n",
    "    print(spname, *data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
